### ISSUE TITLE
Use volatile read in `BaseRevision::is_supported()`

### ISSUE THREAD

**Issue by UditDey** [limine-bootloader/limine-rs#28]
I'm making a kernel using the limine-rs template as reference and noticed that the `_start()` function included in the template doesn't draw anything on the framebuffer:

```rust
#[no_mangle]
unsafe extern "C" fn _start() -> ! {
    assert!(BASE_REVISION.is_supported());

    if let Some(framebuffer_response) = FRAMEBUFFER_REQUEST.get_response() {
        if let Some(framebuffer) = framebuffer_response.framebuffers().next() {
            for i in 0..100_u64 {
                // Calculate the pixel offset using the framebuffer information we obtained above.
                // We skip `i` scanlines (pitch is provided in bytes) and add `i * 4` to skip `i` pixels forward.
                let pixel_offset = i * framebuffer.pitch() + i * 4;

                // Write 0xFFFFFFFF to the provided pixel offset to fill it white.
                *(framebuffer.addr().add(pixel_offset as usize) as *mut u32) = 0xFFFFFFFF;
            }
        }
    }

    hcf();
}
```

Looking at the disassembly:
```
Disassembly of section .text:

ffffffff80000000 <_start>:
ffffffff80000000:       50                      push   rax
ffffffff80000001:       48 8d 3d f0 10 00 00    lea    rdi,[rip+0x10f0]        # ffffffff800010f8 <_ZN36_$LT$T$u20$as$u20$core..any..Any$GT$7type_id17h654c8c39a6067c34E+0x1018>
ffffffff80000008:       48 8d 15 39 21 00 00    lea    rdx,[rip+0x2139]        # ffffffff80002148 <_ZN36_$LT$T$u20$as$u20$core..any..Any$GT$7type_id17h654c8c39a6067c34E+0x2068>
ffffffff8000000f:       be 2e 00 00 00          mov    esi,0x2e
ffffffff80000014:       e8 77 00 00 00          call   ffffffff80000090 <_ZN4core9panicking5panic17h951cafdfee7cec9cE>
ffffffff80000019:       cc                      int3
ffffffff8000001a:       cc                      int3
ffffffff8000001b:       cc                      int3
ffffffff8000001c:       cc                      int3
ffffffff8000001d:       cc                      int3
ffffffff8000001e:       cc                      int3
ffffffff8000001f:       cc                      int3
```

it seems that `_start()` just immediately panics because the compiler thinks that `assert!(BASE_REVISION.is_supported())` will always fail. Removing the assertion fixes this.

The problem seems to be that `BaseRevision::is_supported()` doesn't use a volatile read:
```rust
/// Check whether the given revision is supported by the bootloader.
pub fn is_supported(&self) -> bool {
    (unsafe { *self.revision.get() }) == 0 // Should use read_volatile() instead of *
}
```

and so the compiler assumes that the read will always return 1, because that's whats written in `BaseRevision::new()`. Making it a volatile read fixes this.

**Comment by lylythechosenone** [comment]
I'll fix this in the next release (probably later today). Thanks for the report.

**Comment by lylythechosenone** [comment]
@UditDey is your `BASE_REVISION` static public?

`read_volatile` is not correct here. It is only for use on memory allocated *outside* of the Rust abstract machine. This normal read should just work, as long as the static is marked public (and potentially `no_mangle`). If your static is already public, this is probably a rustc bug.

**Comment by UditDey** [comment]
Well, `read_volatile` is for memory that is read/written outside the abstract machine, not necessarily allocated outside it.

If I've understood correctly, the base revision tag is modified by the bootloader to communicate whether or not the requested revision is supported. And that's what `BaseRevision::is_supported()` seems to be trying to do, read back that field after we've booted to see whether 0 has been written there by the bootloader or not. This is a write from outside the rust abstract machine and absolutely warrants a volatile read.

Even the other tags use volatile reads to get their response. Look at [`Response<T>`](https://github.com/limine-bootloader/limine-rs/blob/master/src/request.rs#L78). This is why the other tags work without making them public/`#[no_mangle]`

But I also experimented with making it public/`#[no_mangle]` and it worked without the volatile read, which is interesting. Maybe `#[no_mangle]` makes rust relax it's assumptions?

In any case, making it a volatile read both removes the need to make it public/`#[no_mangle]` and also makes it consistent with how the other responses are read.

**Comment by lylythechosenone** [comment]
Yes, the others are actually mistaken in using the volatile read. Not making it public is actually UB.

**Comment by UditDey** [comment]
That is surprising, how does that count as UB? I'm curious

**Comment by lylythechosenone** [comment]
Unless the static is public/`no_mangle`, it is considered internal, and assumed to not be modified outside of local code. We can prove that this is UB, because it has caused this miscompilation. A miscompilation can only occur if UB is involved (excluding a compiler bug). Therefore, some UB must have occured. Obviously, the `read` itself is not UB, because `read` only requires that the pointer is valid and the data satisfies the type's validity invariant. Modifying an `UnsafeCell` is not UB, because that is the purpose of an `UnsafeCell`. Therefore, the UB must have been due to the static not being public. Indeed, the miscompilation was caused by the compiler's assumption that because the static was internal, it could not be externally modified.

**Comment by UditDey** [comment]
I'm sorry but I think you're completely incorrect. There was no "miscompilation", I don't see any UB here. The compiler elided away the pointer read because we didn't tell it that the bootloader can swoop in and modify the original value, and the way you tell it that is by using a volatile read.

If you have an internal value that can be modified by something external, you use a volatile read. That is true in C, it is true in rust.

And although it seems that public/'#[no_mangle]' has the same effect, I'm not aware of this property being documented anywhere. If you want to do it correctly, and make sure it stays correct in all future compiler versions, you use a volatile read.

I'm surprised you're confused about this considering that you got this right in `Response<T>` where you do in fact use a volatile read! I have never in my life seen marking something as public/`#[no_mangle]` be touted as equivalent to marking it as volatile.

**Comment by lylythechosenone** [comment]
There was *by definition* a miscompilation. This code should read the value. The fact that it was elided means that either the compiler did something wrong, or you did not uphold your promises to it.

The property of `#[no_mangle]` fixing this issue is related to how it interacts with linkage. Without `#[no_mangle]`, the static is marked as a local symbol. It is only marked as exported with `#[no_mangle]`. See [this discussion](https://internals.rust-lang.org/t/precise-semantics-of-no-mangle/4098) as well as the [ABI](https://doc.rust-lang.org/stable/reference/abi.html) for details on that. This is *not* equivalent to volatile. It is a *requirement* to safely write to the value from outside.

I will admit that this is not a very well-walked path. It is not written in black and white. However, `no_mangle` is intended to work this way. When I initially wrote `Response`, this didn't occur to me at all. I didn't question how a bug like this could happen without UB. But now it is clear that there is UB involved.

**Comment by adavis628** [comment]
There was not a miscompilation. Since `BASE_REVISION` is not referenced anywhere else in the program, the compiler is allowed to remove it. To force the compiler to keep it in the final binary, the `#[used]` attribute can be used (see [here](https://rust-lang.github.io/rfcs/2386-used.html)), and it must be used on all limine requests. I've opened limine-bootloader/limine-rust-template#8 to fix this.

**Comment by lylythechosenone** [comment]
Perhaps miscompilation was the wrong word on my part. I'll look over your PR.

**Comment by UditDey** [comment]
I `objdump | grep`'d the generated elf and here's what I saw:
- Without `#[used]` the `BASE_REVISION` static is removed entirely, and the pointer read in `is_supported()` is elided
- `#[no_mangle]` also works since it implies `#[keep]`
- `read_volatile()` also makes sure the static isn't removed

So, technically only one of these is necessary. However this isn't guaranteed to remain true since, technically, `#[used]` can stop the static from being removed, but not stop the pointer read from being elided.

In fact, read this section from the [RFC that introduced `#[used]`](https://rust-lang.github.io/rfcs/2386-used.html):
> Yet another workaround is to force a volatile load of the vector table in some part of the program, usually before main. The compiler will always keep the vector table in this case but this alternative incurs in the cost of a load operation that will never be optimized away by the compiler.

This section explains why `read_volatile()` has the same effect, and also implies that `#[used]` is not the same as `read_volatile()` and therefore if we just use `#[used]` its technically possible for the static to be kept around but for the pointer read to still be elided.

So basically, I think the best option is to use both `#[used]` and `read_volatile()`, the first ensures that the static isn't removed, and the second ensures that the pointer read is not optimized away

**Comment by lylythechosenone** [comment]
Actually, @UditDey, can you verify whether or not the symbol is considered public? Afaik only the no_mangle will be, but I'm not sure.

**Comment by UditDey** [comment]
If by public you mean exported and visible in the symbol table, then no. It won't be public unless `#[no_mangle]` or `#[export]` are used. But the symbol doesn't need to be public. In case you're confused about it, the bootloader finds the tags by scanning the file for magic numbers, it doesn't refer to the symbol table at all.

**Comment by lylythechosenone** [comment]
Yes, I know that. But if it's not public, llvm still has every right to elide it, and `used` is not perfect.

**Comment by adavis628** [comment]
After rereading the rfc, it seems that `#[used]` appends the static to llvm's [`@llvm.used`](https://llvm.org/docs/LangRef.html#the-llvm-used-global-variable) list. This makes the compiler behave as if there is a reference to the static that it cannot see (effectively making it public), and it will thus keep it in the binary. Since limine requests have interior mutability via `UnsafeCell`, the compiler should also not be able to elide the pointer read.

However, it seems that `#[used]` does not prevent the _linker_ from removing the static if it is not used by any function. I've updated the PR with a comment explaining this.

**Comment by lylythechosenone** [comment]
Alright, then `#[used]` is completely fine. I'll accept your PR.

As for this issue, I'm not sure how we should proceed. The technically correct solution would be to remove all volatile reads, but the one which would cause fewer bugs would be to add a volatile read in `BaseRevision`. Let me know if you have a preference.

**Comment by UditDey** [comment]
> Since limine requests have interior mutability via `UnsafeCell`, the compiler should also not be able to elide the pointer read

This is false. `UnsafeCell` does not ensure that reads to it are never elided. It's only purpose is to enable interior mutability.

Just to make my point, compare the output assembly on [this](https://godbolt.org/z/vGzo711nG) and [this](https://godbolt.org/z/Exs5W4E5f). The one with `read_volatile()` actually does the read, whereas the one without it elides it, despite both using `UnsafeCell`. It is absolutely not a rule of the language that a pointer originating from an `UnsafeCell` behaves any different than any other pointer. If you want to prevent the read elision, you use `read_volatile()`.

> The technically correct solution would be to remove all volatile reads

Again, I don't understand where this is coming from. Unless the rust docs actually say that using `#[keep]` ensures no pointer read elision, the only fool proof way to prevent that, in every future rust compiler version, is to use what is actually documented and guaranteed to work: `read_volatile()`

And even if you're fine with relying on undocumented behavior, how do you conclude the correct solution is to _remove_ whatever volatile reads there are??

I am upset because what I thought was going to be a one-line fix has turned into a discussion of the semantics of compiler optimizations :(

I'll reiterate: The best solution is to both  use `#[keep]` on all our request statics, and `read_volatile()` for every response read

**Comment by adavis628** [comment]
The reason why the compiler was able to elide the pointer read in your example was because the compiler could (correctly) infer that the value could not have changed by the time it was read. However, marking a static with `#[used]` forces the compiler/llvm to assume some other unknown part of the code references it. However, if the static could not change, the compiler could still elide it (see [here](https://godbolt.org/z/cMMEeMfaz)). But since `UnsafeCell` provides interior mutability, the compiler can no longer make that assumption.

Imo, volatile reads are not necessary and could prevent correct optimizations such as reordering reads. The need to use `#[used]` should be documented though.

**Comment by TobiasSN** [comment]
I don't imagine checking the base revision of the boot protocol is a particularly time-sensitive operation. Volatile reads seems more practical for end users than `#[used]`, and besides, Limine already expects you to use `volatile` in C.

**Comment by lylythechosenone** [comment]
Alright, I suppose I'll just add the volatile then. I'd still like that PR to land, however.

**Comment by UditDey** [comment]
> However, marking a static with #[used] forces the compiler/llvm to assume some other unknown part of the code references it

If you can find the part of the rust docs that states that `#[used]` suppresses pointer optimizations I'll agree right away. However it's stated purpose is to just ensure that the item is not removed. The fact that it also suppresses pointer optimization is incidental, undocumented behavior that is not guaranteed to remain.

Look, I'm not being cranky for no reason here. I've been bitten by things like this!! I've written low level rust code that worked fine, I run `rustup update` and it stops working, because turns out that all along I was relying on incidental behavior instead of following the stated rules.

And as @TobiasSN points out, you literally need the `volatile` marker in C too!! Would you insist that instead of making it `volatile` you should `__declspec(export)` it?

And yes @lylythechosenone, that PR should land. `#[keep]` is a good idea (and so is `read_volatile()` :P)

**Comment by adavis628** [comment]
> If you can find the part of the rust docs that states that #[used] suppresses pointer optimizations I'll agree right away.

While there is nothing that explicitly states this behavior, it follows logically from the given conditions. According to [the rfc](https://rust-lang.github.io/rfcs/2386-used.html), 

> The `#[used]` attribute can only be used on static variables. Static variables marked with this attribute will be appended to the special `@llvm.used` global variable when lowered to LLVM IR.

This tells rust/llvm that the static is reachable from some unknown outside code (by limine in this case). Also, the static is mutable via [`UnsafeCell`](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html). Since the unknown code can mutate the static, it follows that the compiler will not be able to make any assumptions about its value, preventing inlining. To me, it seems that this behavior is not incidental; rather, it follows logically from the condition that it can be mutated by unknown code.

> And as @TobiasSN points out, you literally need the volatile marker in C too!! Would you insist that instead of making it volatile you should __declspec(export) it?

In the [limine C template](https://github.com/limine-bootloader/limine-c-template), removing `volatile` from
```C
static volatile LIMINE_BASE_REVISION(1);
```
does indeed break the example. However, replacing it with `__attribute((used))` ([the C equivalent of `#[used]`](https://llvm.org/docs/LangRef.html#the-llvm-used-global-variable)) fixes it. Perhaps the C example could be updated to use this in the future.

> I don't imagine checking the base revision of the boot protocol is a particularly performance-sensitive operation.

I agree, but there is a significant difference in generated code when the read is able to be hoisted out of a loop (https://godbolt.org/z/7xE6PjGM7). Granted, this should occur rarely, if at all, in real code.

In my opinion, the biggest reason for volatile operations is that it would mitigate the effects of someone forgetting to mark statics with `#[used]`. However, I think that documenting why limine requests need to be marked with `#[used]` along with being referenced in a function should prevent this from happening. Perhaps the limine utils could eventually have the capability to warn against a missing base revision tag.

**Comment by lylythechosenone** [comment]
> Perhaps the limine utils could eventually have the capability to warn against a missing base revision tag.

This is a decent idea. I'll bring it up to mint.
